### PR TITLE
Render arbitrarily nested lists in the UI

### PR DESCRIPTION
**Pull request by jmchilton**
I started with #5013 and simply refactored the client until I eliminated 20 classes (Backbone models and views) and many places collection type was dispatched on to produce different classes without different behavior. This resulted in no features lost but now arbitrarily nested lists render properly.

The classes lost in this refactoring include:

- ListDatasetCollection
- PairDatasetCollection
- NestedPairDCDCE
- NestedPairDCDCECollection
- ListPairedDatasetCollection
- NestedListDCDCE
- NestedListDCDCECollection
- ListOfListsDatasetCollection
- ListCollectionViewEdit
- PairCollectionViewEdit
- ListOfPairsCollectionViewEdit
- ListOfListsCollectionViewEdit
- ListCollectionView
- PairCollectionView
- ListOfPairsCollectionView
- ListOfListsCollectionView
- HistoryPairDatasetCollection
- HistoryListPairedDatasetCollection
- HistoryListOfListsDatasetCollection
- NestedCollectionElementViewEdit

The exports for various modules this touches reveal how this simplified the interface:

```diff
export default {
- ListDatasetCollection: ListDatasetCollection,
- PairDatasetCollection: PairDatasetCollection,
- ListPairedDatasetCollection: ListPairedDatasetCollection,
- ListOfListsDatasetCollection: ListOfListsDatasetCollection
+ DatasetCollection: DatasetCollection,
};
```

```diff
 export default {
     CollectionViewEdit: CollectionViewEdit,
-    ListCollectionViewEdit: ListCollectionViewEdit,
-    PairCollectionViewEdit: PairCollectionViewEdit,
-    ListOfPairsCollectionViewEdit: ListOfPairsCollectionViewEdit,
-    ListOfListsCollectionViewEdit: ListOfListsCollectionViewEdit
 };
```

```diff
 export default {
     CollectionView: CollectionView,
-    ListCollectionView: ListCollectionView,
-    PairCollectionView: PairCollectionView,
-    ListOfPairsCollectionView: ListOfPairsCollectionView,
-    ListOfListsCollectionView: ListOfListsCollectionView
 };
```

```diff
 export default {
-    HistoryListDatasetCollection: HistoryListDatasetCollection,
-    HistoryPairDatasetCollection: HistoryPairDatasetCollection,
-    HistoryListPairedDatasetCollection: HistoryListPairedDatasetCollection,
-    HistoryListOfListsDatasetCollection: HistoryListOfListsDatasetCollection
+    HistoryDatasetCollection: HistoryDatasetCollection,
 };
```

This also builds on the collection state tests from #5013 with many more collection tests - testing the rendering of various kinds of collections and some collection history operations (e.g. nametags).

### CODE REVIEW

**Comment by dannon** [comment]
Haven't tested this yet, but *thank you* for flattening this.

**Comment by jmchilton** [comment]
That test failing Selenium test passes pretty consistently locally and fails pretty consistently on Jenkins - huh. How odd - I'll mark this as WIP.

File: client/galaxy/scripts/mvc/collection/collection-model.js [hunk 1]
```diff
@@ -398,18 +398,18 @@ var NestedListDCDCECollection = NestedDCDCECollection.extend({
 
 //==============================================================================
 /** @class Backbone Model for a DatasetCollection (list) that contains other lists. */
-var ListOfListsDatasetCollection = DatasetCollection.extend({
+var NestedDatasetCollection = DatasetCollection.extend({
```
Comment (martenson) [inline client/galaxy/scripts/mvc/collection/collection-model.js]: This makes so much sense.

**Review by martenson — APPROVED** [review]
https://twitter.com/martenson/status/932646373949878275

**Pull request merged**